CMake: remove MBED_PATH and use <PROJECT-NAME>_SOURCE_DIR/MODULE_PATH instead

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

This PR is helping UNITTESTS so it should be merged prior #14426 - they need to use files within Mbed OS and already defines `project()`.

We used MBED_PATH previously, it's own defined variable for Mbed OS project root. Usage were to reference CMake scripts or files out of it's current source tree. 

The approach should be to use cmake module path for listfiles like .cmake to find them. Note, due to our target labels, some CMake files are hidden so `include` must be also protected if it touches conditionally added library (for instance TFM).

For CMakeLists.txt files, use `mbed-os_SOURCE_DIR` - this is useful for CMakeLists.txt. For listfiles like .cmake files, please use CMAKE_CURRENT_LIST_DIR (as they are included).


#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 committed:
CMake: use project name for Mbed OS

This enables multiple variables we can use. One of them is source dir for a project.

It will replace MBED_PATH in the tree.

0xc0170 committed:
CMake targets: remove MBED_PATH, use stadard variables

- list files included via module path
- <project-name>_SOURCE_DIR for sources that are out of the current processed CMake
- CMAKE_CURRENT_LIST_DIR for listfiles

0xc0170 committed:
CMake platform: remove MBED_PATH, use standard variables

As we still use target labels, TFM is not visible to CMake. Protect include of their scripts with the same mechanism.

0xc0170 committed:
CMake tools: use CMAKE_CURRENT_LIST_DIR instead of MBED_PATH

Excluding greentea tests, they will be fixed separately (more work needed).

0xc0170 committed:
CMake app: add our cmake folder to CMAKE_MODULE_PATH

All files within tools/cmake can now include within tools/cmake. We do not expose
internal folders like core/profiles/toolchains.
They should be included via toolchain file (mbed_toolchain in our case).

rajkan01 commented on connectivity/FEATURE_BLE/source/cordio/TESTS/cordio_hci/driver/CMakeLists.txt:
```diff
@@ -3,10 +3,10 @@
 cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
 
-set(MBED_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../../../../../../.. CACHE INTERNAL "")
+set(mbed-os_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/../../../../../../.. CACHE INTERNAL "")
```
As a current implementation, every greentea test has its CMakeLists.txt similar to the application CMake. Why do we need to change into `mbed-os_SOURCE_DIR`? this will be a problem when we do "Separate GreenTea from Mbed"

0xc0170 replied:
I see, this is setting it. I missed it during refactoring, I'll remove all

rajkan01 commented on connectivity/FEATURE_BLE/source/cordio/TESTS/cordio_hci/driver/CMakeLists.txt:
```diff
@@ -3,10 +3,10 @@
 set(TEST_TARGET mbed-connectivity-ble-cordio-hci-driver)
 
-include(${MBED_PATH}/tools/cmake/mbed_greentea.cmake)
+include(${mbed-os_SOURCE_DIR}/tools/cmake/mbed_greentea.cmake)
```
Sorry I missed to comment here, Do we need to change this as well because if it is top-level CMake similar to application one then PROJECT_NAME and CMAKE_PROJECT_NAME will be `mbed-connectivity-ble-cordio-hci-driver` and the `mbed-os_SOURCE_DIR` variable is not yet created

0xc0170 replied:
The test should not care where it is.

How we create greentea shall be changed and this MBED_PATH should be fixed there. I'll create a separate issue for greentea tests, how they use MBED_PATH and get mbed-os in. I'll revert the changes here.

I'll keep this to fix the Mbed OS files (excluding greentea tests).

0xc0170 replied:
Created https://github.com/ARMmbed/mbed-os/issues/14523 and I'll revert now greentea tests changes.

rajkan01 commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -3,7 +3,7 @@
 set(MBED_CONFIG_PATH ${CMAKE_CURRENT_BINARY_DIR} CACHE INTERNAL "")
 
-include(${MBED_PATH}/tools/cmake/app.cmake)
+include(${mbed-os_SOURCE_DIR}/tools/cmake/app.cmake)
```
Can we fix this as part of #14523? and revert this change

0xc0170 replied:
Updated

rajkan01 reviewed (dismissed):
LGTM

rwalton-arm commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_B1/CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
-include(${MBED_PATH}/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts/mbed_set_post_build_tfm.cmake)
+include(${mbed-os_SOURCE_DIR}/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts/mbed_set_post_build_tfm.cmake)
```
I'm wondering if we can use [CMAKE_MODULE_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_MODULE_PATH.html?highlight=cmake_module_path) to avoid hard coding all of the include paths. We could set the module directories in `mbed-os/CMakeLists` and then include modules with e.g `include(mbed_set_post_build_tfm)` in other files. The benefit of this approach, in my view, is if we ever move the location of these files we will have fewer changes to make.

0xc0170 replied:
+1, I'll check how we can create the modules for our scripts

rwalton-arm commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts/mbed_set_post_build_tfm.cmake:
```diff
@@ -17,7 +17,7 @@ function(mbed_post_build_tfm_sign_image
     set(post_build_command
         COMMAND ${Python3_EXECUTABLE}
-            ${MBED_PATH}/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts/generate_mbed_image.py
+            ${mbed-os_SOURCE_DIR}/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts/generate_mbed_image.py
```
Can this not be reduced to `${CMAKE_CURRENT_LIST_DIR}/generate_mbed_image.py`?

0xc0170 replied:
we are in the targets area, not in the tools so current list dir points to targets/ somewhere.

0xc0170 replied:
Anyway, this path is incorrect. I'll fix it.

rwalton-arm commented on tools/cmake/app.cmake:
```diff
@@ -12,11 +12,11 @@ include(${MBED_CONFIG_PATH}/mbed_config.cmake)
 if(NOT CMAKE_TOOLCHAIN_FILE OR MBED_TOOLCHAIN_FILE_USED)
     set(MBED_TOOLCHAIN_FILE_USED TRUE CACHE INTERNAL "")
-    include(${MBED_PATH}/tools/cmake/toolchain.cmake)
+    include(${mbed-os_SOURCE_DIR}/tools/cmake/toolchain.cmake)
```
`${CMAKE_CURRENT_LIST_DIR}/toolchain.cmake`?

0xc0170 replied:
yes, fixed in the later commit.

rwalton-arm commented on tools/cmake/app.cmake:
```diff
@@ -12,11 +12,11 @@ include(${MBED_CONFIG_PATH}/mbed_config.cmake)
 
 # Specify available build profiles and add options for the selected build profile
-include(${MBED_PATH}/tools/cmake/profile.cmake)
+include(${mbed-os_SOURCE_DIR}/tools/cmake/profile.cmake)
```
`${CMAKE_CURRENT_LIST_DIR}/profile.cmake`?

0xc0170 replied:
yes, fixed in the later commit.

rwalton-arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -3,7 +3,7 @@
 set(MBED_CONFIG_PATH ${CMAKE_CURRENT_BINARY_DIR} CACHE INTERNAL "")
 
-include(${MBED_PATH}/tools/cmake/app.cmake)
+include(${mbed-os_SOURCE_DIR}/tools/cmake/app.cmake)
```
`${CMAKE_CURRENT_LIST_DIR}/app.cmake`?

0xc0170 replied:
yes, fixed in the later commit. I realized we should use list dir in tools

rwalton-arm commented on tools/cmake/profile.cmake:
```diff
@@ -29,4 +29,4 @@ else()
 endif()
 
-include(${MBED_PATH}/tools/cmake/profiles/${LOWERCASE_CMAKE_BUILD_TYPE}.cmake)
+include(${mbed-os_SOURCE_DIR}/tools/cmake/profiles/${LOWERCASE_CMAKE_BUILD_TYPE}.cmake)
```
`${CMAKE_CURRENT_LIST_DIR}/profiles/${LOWERCASE_CMAKE_BUILD_TYPE}.cmake`?

rwalton-arm commented on tools/cmake/app.cmake:
```diff
@@ -12,11 +12,11 @@ include(${MBED_CONFIG_PATH}/mbed_config.cmake)
 if(NOT CMAKE_TOOLCHAIN_FILE OR MBED_TOOLCHAIN_FILE_USED)
     set(MBED_TOOLCHAIN_FILE_USED TRUE CACHE INTERNAL "")
-    include(${mbed-os_SOURCE_DIR}/tools/cmake/toolchain.cmake)
+    include(${CMAKE_CURRENT_LIST_DIR}/toolchain.cmake)
```
OK, I see you changed these in a later commit. It would be easier to review if we didn't make changes to already reviewed commits later on in the PR.

0xc0170 replied:
I do not want to push force while it's being reviewed but I can clean it up.

0xc0170 replied:
> It would be easier to review if we didn't make changes to already reviewed commits later on in the PR.

I do not follow. I fixed earlier commit in the subsequent one. how else can I make review easier without rebasing?

rwalton-arm replied:
I usually would force push, but marking them as `fixup` commits would also work for me.

0xc0170 replied:
+1 for force push but that is hard for some reviewers so I can mark as fixu and later rebase to get them out of history.

0xc0170 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_B1/CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
-include(${MBED_PATH}/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts/mbed_set_post_build_tfm.cmake)
+include(mbed_set_post_build_tfm)
```
this fails to build as this target might be processed prior TFM (where the scripts are) and include fails to find it.
The problem is we moved scripts into libraries and libraries can be loaded later. `mbed_set_post_build` - this works as the module path is updated prior adding any additional Mbed OS library.

There are dependencies between targets and libraries like tfm and others .

0xc0170 replied:
Deps like this should be resolved as we are loading first generic modules, target as the last. Checking why it can't find it.

0xc0170 replied:
The error I get now:
```
CMake Error at mbed-os/targets/TARGET_ARM_SSG/TARGET_MUSCA_B1/CMakeLists.txt:4 (include):
  include could not find load file:

    mbed_set_post_build_tfm
```

Update: found a reason, see below

0xc0170 committed:
CMake: fix for CMAKE_CURRENT_LIST_DIR in functions

CMAKE_CURRENT_LIST_DIR behaves differently in functions. We store it in the CMakeLists itself, so anyone
calling a function would get the actual list dir where the scripts are.

To illustrate: if I call a function from src/CMakelists.txt, function located in src/scripts, `CMAKE_CURRENT_LIST_DIR` in the function would point
to the src/ folder but not to src/scripts.

0xc0170 committed:
CMake: add mbed_ prefix to our CMake files

To avoid conflicts as we expose our CMake list files via CMAKE_MODULE_PATH. Some files already include it, this
aligns the rest of files.

I leave app.cmake as it is - it's user facing and it would be breaking change. We can clean this one for the next major version.

rwalton-arm commented on targets/TARGET_NXP/TARGET_MCUXpresso_MCUS/TARGET_LPC54114/device/TARGET_LPC54114_M4/CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
-include(${MBED_PATH}/targets/TARGET_NXP/scripts/mbed_set_post_build_nxp.cmake)
+include(${mbed-os_SOURCE_DIR}/targets/TARGET_NXP/scripts/mbed_set_post_build_nxp.cmake)
```
Did we miss this one?

0xc0170 replied:
yes, let me search again entgire tree if I've missed any other

0xc0170 replied:
fixed

rwalton-arm reviewed (commented):
The changes look good to me. I think we might have missed one include path.

0xc0170 committed:
CMake: set CMAKE_MODULE_PATH in the Mbed OS CMakelists.txt

Update CMAKE_MODULE_PATH at once place.

Note, we update also CMAKE_MODULE_PATH in app.cmake. This is temporary until we get a proper way to include
Mbed Os (removing app.cmake need to be included by an application).
